mgr/volumes: versioned subvolume provisioning

vshankar opened the pull request:
work in progress, but should be good for initial review. I've tried to separate out commits for easier review (couldn't sanely split it into separate PRs).

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,100 @@
+        except (VolumeException, MetadataMgrException, cephfs.Error) as e:
+            try:
+                # cleanup subvol path on best effort basis
+                # TODO: this *does not* work now since there are subdirs -- FIXME.
```
better -- w/ the changes that I have locally, the subvolume is moved to trash and purge thread takes care of deleting.

vshankar commented on src/pybind/mgr/volumes/fs/operations/group.py:
```diff
@@ -0,0 +1,163 @@
+                return []
+            raise
+
+def create_group(fs, vol_spec, groupname, pool, mode, uid, gid):
```
Having these helpers as a part of the class and return an instance of the class does not do much help.

vshankar replied:
that's a quirk in the design since subvolumes are versioned. maybe, the group could have been created with a fixed version and that would have kept things similar to subvolumes. I'm inclined to do that later if you agree.

vshankar replied:
`create_group()` can additionally yield a group object (just like `open(2)`), but I did not find a case (workflow) where one would create the group and start using it right away (not as two separate manager calls).

vshankar commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -0,0 +1,129 @@
+        except cephfs.Error as e:
+            raise VolumeException(-e.args[0], e.args[1])
+
+def create_trashcan(fs, vol_spec):
```
ditto -- as per https://github.com/ceph/ceph/pull/31763/files/756cd21e84d82048875a3758c84d3bc36cece329..72a2135dd49daf8b620795de11ec435f02c2ba3e#diff-64535a5a1f7ccd5c510701022b466222

vshankar commented on src/pybind/mgr/volumes/fs/vol_spec.py:
```diff
@@ -0,0 +1,37 @@
+    DEFAULT_NS_PREFIX = "fsvolumens_"
+
+    def __init__(self, snapshot_prefix, subvolume_prefix=None, pool_ns_prefix=None):
+        self.snapshot_prefix = snapshot_prefix
```
so as to not pass rados instance around.

vshankar commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -0,0 +1,88 @@
+        orchestrator.raise_if_exception(completion)
+    except (ImportError, orchestrator.OrchestratorError):
+        return 0, "", "Volume created successfully (no MDS daemons created)"
+    except Exception as e:
```
yeh -- thanks!

vshankar replied:
guess I didn't want logger to be imported in a util module.

vshankar commented on src/pybind/mgr/volumes/fs/operations/group.py:
```diff
@@ -0,0 +1,163 @@
+        st = fs.stat(group.path)
+        group.uid = int(st.st_uid)
+        group.gid = int(st.st_gid)
+    except cephfs.ObjectNotFound:
```
makes sense -- thanks!

vshankar commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -0,0 +1,88 @@
+        return 0, "", "Volume created successfully (no MDS daemons created)"
+    except Exception as e:
+        return -errno.EINVAL, "", str(e)
+    return 0, "", ""
```
I thought of doing that, but then, this was kept as it is as for the error cases we return the error string too.

vshankar commented:
this passes volumes test in teuthology -- marking the pr as ready for review.

addressed comments by @ajarr and rebased. please take a look.

batrick commented:
@vshankar please rebase and linka  teuthology run. I'll take a look Thrusday or Friday and merge.

vshankar commented:
> @vshankar please rebase and linka teuthology run. I'll take a look Thrusday or Friday and merge.

I'll do this tomorrow.

vshankar commented:
> I'll do this tomorrow.

@batrick -- I'm adding the auto-upgrade (for auto-generating a metadata config for existing subvolumes) functionality atm. will post when done.

vshankar commented:
http://pulpito.ceph.com/vshankar-2019-12-17_09:07:40-fs-wip-versioned-mgr-volumes-testing-basic-mira/

vshankar commented:
> http://pulpito.ceph.com/vshankar-2019-12-17_09:07:40-fs-wip-versioned-mgr-volumes-testing-basic-mira/

minor issue with rebase -- rerunning.

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1193,3 +1193,30 @@ def test_mgr_eviction(self):
+
+        # emulate a old-fashioned subvolume
+        createpath = os.path.join(".", "volumes", "_nogroup", subvolume)
+        self.mount_a.run_shell(['mkdir', '-p', createpath])
```
Let's have one for a subvolume with a real group name too!

vshankar replied:
ACK

vshankar commented:
fixed and updated

@batrick -- http://pulpito.ceph.com/vshankar-2019-12-18_11:47:27-fs-wip-versioned-mgr-volumes-testing-basic-smithi/ (the "evicting unresponsive client" is not ceph-mgr).

vshankar commented:
(minor update to the test)

vshankar commented:
> So whenever mgr/volumes implements the access management to subvolumes like the ceph_volume_client, it can continue using the above metadata file format. And this shouldn't interfere with your design of subvolume cloning, right?

yeh -- this will not interfere with the access management policy you described.

vshankar commented:
@batrick @ajarr @joscollin rebased and updated -- please take a look

vshankar commented:
http://pulpito.ceph.com/vshankar-2020-01-06_13:33:25-fs-wip-versioned-mgr-volumes-distro-basic-mira/

batrick commented:
Needs rebased

vshankar commented:
> Needs rebased

waiting for @ajarr review

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1193,3 +1193,44 @@ def test_mgr_eviction(self):
+        self.assertNotEqual(subvolpath2, None)
+        subvolpath2 = subvolpath2.rstrip() # remove "/" prefix and any trailing newline
+
+        # and... the subvolume path returned should be what we created behind the scene
```
that's because upgrade _nneds_ to be transparent -- we could additionally check contents of `_legacy` (to have just a slight bit of confidence), but it does not ensure that the subvolume path is picked up from there too ;)

ajarr replied:
Yes, in addition, would be better. Anyways, this seems good enough as you say.

vshankar commented:
@ajarr I'll push an update (w/ rebase) later today (waiting for more comments if any).

ajarr commented:
> @ajarr I'll push an update (w/ rebase) later today (waiting for more comments if any).

@vshankar LGTM. Tested locally and works fine. Please rebase.

vshankar commented:
> @vshankar LGTM. Tested locally and works fine. Please rebase.

Thanks for reviewing. rebased and updated.

batrick commented:
@vshankar I'm relying on @ajarr to pick through the code here but I did notice that `SubVolume` from `src/pybind/mgr/volumes/fs/subvolume.py` is no longer used? Is that stale code that needs removed?

batrick commented:
> @vshankar I'm relying on @ajarr to pick through the code here but I did notice that `SubVolume` from `src/pybind/mgr/volumes/fs/subvolume.py` is no longer used? Is that stale code that needs removed?

With this said, I'm good to merge this but I do wonder if there is any other stale code that also needs removed.

vshankar commented:
> With this said, I'm good to merge this but I do wonder if there is any other stale code that also needs removed.

I *might* have just missed removing that -- I'll send that as a follow up.
